compiler/rustc_mir_dataflow/src/value_analysis.rs

```diff
@@ -0,0 +1,937 @@
+    ///
+    /// This is currently the only way to create a [`Map`]. The way in which the tracked places are
+    /// chosen is an implementation detail and may not be relied upon (other than that their type
+    /// passes the filter).
+    #[instrument(skip_all, level = "debug")]
+    pub fn from_filter<'tcx>(
+        tcx: TyCtxt<'tcx>,
+        body: &Body<'tcx>,
+        filter: impl FnMut(Ty<'tcx>) -> bool,
+    ) -> Self {
+        let mut map = Self::new();
+        map.register_with_filter(tcx, body, filter, &escaped_places(body));
+        debug!("registered {} places ({} nodes in total)", map.value_count, map.places.len());
+        map
+    }
+
+    /// Register all non-excluded places that pass the filter.
+    fn register_with_filter<'tcx>(
+        &mut self,
+        tcx: TyCtxt<'tcx>,
+        body: &Body<'tcx>,
+        mut filter: impl FnMut(Ty<'tcx>) -> bool,
+        exclude: &FxHashSet<Place<'tcx>>,
+    ) {
+        // We use this vector as stack, pushing and popping projections.
+        let mut projection = Vec::new();
+        for (local, decl) in body.local_decls.iter_enumerated() {
+            self.register_with_filter_rec(
+                tcx,
+                local,
+                &mut projection,
+                decl.ty,
+                &mut filter,
+                exclude,
+            );
+        }
+    }
+
+    /// Register fields of the given (local, projection) place.
+    ///
+    /// Invariant: The projection must only contain fields.
+    fn register_with_filter_rec<'tcx>(
+        &mut self,
+        tcx: TyCtxt<'tcx>,
+        local: Local,
+        projection: &mut Vec<PlaceElem<'tcx>>,
+        ty: Ty<'tcx>,
+        filter: &mut impl FnMut(Ty<'tcx>) -> bool,
+        exclude: &FxHashSet<Place<'tcx>>,
+    ) {
+        if exclude.contains(&Place { local, projection: tcx.intern_place_elems(projection) }) {
+            // This will also exclude all projections of the excluded place.
+            return;
+        }
+
+        // Note: The framework supports only scalars for now.
+        if filter(ty) && ty.is_scalar() {
+            // We know that the projection only contains trackable elements.
+            let place = self.make_place(local, projection).unwrap();
+
+            // Allocate a value slot if it doesn't have one.
+            if self.places[place].value_index.is_none() {
+                self.places[place].value_index = Some(self.value_count.into());
+                self.value_count += 1;
+            }
+        }
+
+        // Recurse with all fields of this place.
+        iter_fields(ty, tcx, |variant, field, ty| {
+            if variant.is_some() {
+                // Downcasts are currently not supported.
+                return;
+            }
+            projection.push(PlaceElem::Field(field, ty));
+            self.register_with_filter_rec(tcx, local, projection, ty, filter, exclude);
+            projection.pop();
+        });
```

Comment:
Why are we recursively iterating over fields? This only supports scalars, right? And scalars don't have fields?

Reply:
It only supports scalars in the sense that only scalars are tracked. If you have `struct S(i32, u32)`, both of its fields can be tracked. Another way to look at this is that we currently support place expressions which have scalar types and only field projections, e.g., `_1.2.0.2: usize`. Previously, derefs and downcasts were allowed as well, as in `*((*_1).0 as X).2: f32`, but both of these need more work (and deref might not be feasible for now).